neo4j/src/main/java/org/apache/metamodel/neo4j/utils/ColumnTypeResolver.java

```diff
@@ -0,0 +1,126 @@
+        } catch (final JSONException e) {
+            // ignore missing data
+        }
+
+        fillColumnTypesFromRemainingColumns(columnNames, columnTypes);
+        return columnTypes.toArray(new ColumnType[columnTypes.size()]);
+    }
+
+    private void fillColumnTypesFromData(final JSONObject jsonObject, final List<String> columnNames,
+            final List<ColumnType> columnTypes) throws JSONException {
+        final String dataKey = "data";
+
+        if (jsonObject.has(dataKey)) {
+            final JSONObject data = jsonObject.getJSONObject(dataKey);
+            final Iterator<?> keysIterator = data.keys();
+
+            while (keysIterator.hasNext()) {
+                final String key = (String) keysIterator.next();
+                final ColumnType type = getTypeFromValue(data, key);
+                columnTypes.add(type);
+                removeIfAvailable(columnNames, key);
+            }
+        }
+    }
+
+    private void fillColumnTypesFromMetadata(final JSONObject jsonObject, final List<String> columnNames,
+            final List<ColumnType> columnTypes) throws JSONException {
+        final String metadataKey = "metadata";
+
+        if (jsonObject.has(metadataKey)) {
+            final JSONObject metadata = jsonObject.getJSONObject(metadataKey);
+
+            if (metadata.has("id")) {
+                columnTypes.add(ColumnType.BIGINT);
+                removeIfAvailable(columnNames, "_id");
+            }
+        }
+    }
+
+    private void fillColumnTypesFromRemainingColumns(final List<String> columnNames,
+            final List<ColumnType> columnTypes) {
+        for (final String remainingColumnName : columnNames) {
+            if (remainingColumnName.contains("rel_")) {
+                if (remainingColumnName.contains("#")) {
+                    columnTypes.add(ColumnType.ARRAY);
+                } else {
+                    columnTypes.add(ColumnType.BIGINT);
+                }
+            } else {
+                columnTypes.add(ColumnType.STRING);
+            }
+        }
+    }
+
+    private void removeIfAvailable(final List<String> list, final String key) {
+        if (list.contains(key)) {
+            list.remove(key);
+        }
+    }
+
+    private ColumnType getTypeFromValue(final JSONObject data, final String key) {
+        try {
+            final Class<? extends Object> keyClass = data.get(key).getClass();
+            
+            if (keyClass.equals(Boolean.class)) {
+                return ColumnType.BOOLEAN;
+            } else if (keyClass.equals(Integer.class)) {
+                return ColumnType.INTEGER;
+            } else if (keyClass.equals(Long.class)) {
+                return ColumnType.BIGINT;
+            } else if (keyClass.equals(Double.class)) {
+                return ColumnType.DOUBLE;
+            } else if (keyClass.equals(JSONArray.class)) {
+                return ColumnType.ARRAY;
+            } else if (keyClass.equals(JSONObject.class)) {
```

Comment:
In practice a property will never be a JSONObject (or MAP), because Neo4j informed me that `Neo.ClientError.Statement.TypeError: Property values can only be of primitive types or arrays thereof`, so I think we can remove this option.